### ISSUE TITLE
Log rotation

### ISSUE THREAD

**Issue by maddinat0r** [maddinat0r/samp-log-core#8]
That's definitely a thing I should implement.

**Comment by mikroware** [comment]
Is there any prospect on this feature? The only thing it should probably do is dynamically compute the log file name based on a datetime format.

Do you perhaps know some other way of doing this on a linux machine? Like a daily rename script. Will renaming not crash the plugin while it is running?

**Comment by maddinat0r** [comment]
I can't tell when exactly I'm going to work on this, but finishing the log-core has a high priority on my todo list.

You can create a cronjob to rename/move/delete the log file. The log-core doesn't keep the log file open, it opens and closes it on every log write. So technically there is a small chance that you could rename it while the log-core writes into the log file. However this isn't really an issue on Linux, as doing those actions at the same time should be safe (see https://superuser.com/a/251244).

**Comment by mikroware** [comment]
Thanks for your quick answer! I will keep an eye on here for any updates.

Good to know that renaming/moving on Linux should not be an issue (had such issues when running under windows long ago). Going to make my rotation with linux logrotate now.

**Comment by maddinat0r** [comment]
File size based log rotation is now implemented (8f5788269c8fca637ce3ec31ecebe9c81194a367), and configurable like this:
```yaml
Logger:
  plugins/mysql:
    LogLevel: All
    LogRotation:
      Type: Size
      Trigger: 1MB
      BackupCount: 15
```
Size can be specified in kilobytes (KB), megabytes (MB) and gigabytes (GB). Backup count is the number of files to keep, e.g. if there are 15 log files and the logs rotate, the last file is deleted to not have more than 15 rotated logs available.

The only thing missing now is time-based log rotation, which I initially planned to implement by specifying a duration after which the rotation happens. However, this would depend on the time the server started, so specifying a 24 hours duration and starting the server at 6 o'clock would always rotate at 6 o'clock every day.
I'll probably change that to simply specifying if rotation should happen daily, weekly or monthly. This is not very flexible, but easy to configure and implement. Otherwise I'll have to think of a proper syntax that allows specifying any date and duration (like the syntax for Linux cronjobs for example).

**Comment by maddinat0r** [comment]
Implemented date-based log rotation as stated in the above comment:
```yaml
Logger:
  plugins/mysql:
    LogLevel: All
    LogRotation:
      Type: Date
      Trigger: Daily
      BackupCount: 25
```
